Dev/issue 6596 aip restore v3

mcantelon opened the pull request.

mcantelon committed:
Add AIP recovery feature

mistydemeo commented on storage_service/administration/forms.py:
```diff
@@ -104,6 +107,14 @@ class DefaultLocationsForm(SettingsForm):
+        required=False,
+        label="Default AIP recovery locations for new pipelines:"
+        )
+    new_backlog = DefaultLocationField(
```
Wrong variable name? This is shadowing the previous definition of `new_backlog`.

mistydemeo commented on storage_service/administration/forms.py:
```diff
@@ -78,6 +78,9 @@ class CommonSettingsForm(SettingsForm):
     """ Configures common or generic settings that don't belong elsewhere. """
     pipelines_disabled = forms.BooleanField(required=False,
         label="Pipelines are disabled upon creation?")
+    recover_location_uuid = forms.CharField()
```
Instead of requiring the user to set a UUID, how about making the AIP recovery location per-pipeline, and setting it as a default the way that transfer source, AIP storage, etc. gets set below?

mistydemeo commented on storage_service/locations/api/resources.py:
```diff
@@ -449,33 +450,25 @@ def delete_aip_request(self, request, bundle, **kwargs):
+        package = bundle.obj
+        if package.package_type not in Package.PACKAGE_TYPE_CAN_RECOVER:
+            # Can only request recovery of AIPs
+            return http.HttpMethodNotAllowed()
```
Include a message here?

mistydemeo commented on storage_service/locations/models.py:
```diff
@@ -1244,6 +1248,83 @@ def _update_quotas(self, space, location):
+        # Copy corrupt files to storage service staging
+        backup_path = utils.get_setting('recover_backup_path_within_location')
+
+        if backup_path == None:
```
`is None`

mistydemeo commented on storage_service/locations/models.py:
```diff
@@ -1244,6 +1248,83 @@ def _update_quotas(self, space, location):
+                self.current_path)
+            destination_path = os.path.join(
+                origin_location.relative_path,
+                utils.get_setting('recover_backup_path_within_location'))
```
Use the variable assigned above?

mistydemeo commented on storage_service/locations/views.py:
```diff
@@ -44,9 +45,49 @@ def package_list(request):
     return render(request, 'locations/package_list.html', locals())
 
+def aip_recover_request(request):
+    if request.method == 'GET' and utils.get_setting('recover_backup_path_within_location') == None:
```
`is None` here too

mistydemeo commented on storage_service/locations/views.py:
```diff
@@ -44,9 +45,49 @@ def package_list(request):
+        messages.error(request, 'Recover backup path not set.')
+
+    def execution_logic(aip): 
+        recover_location = LocationPipeline.objects.filter(
```
Do you expect more than one recovery location to be defined for a pipeline?

mistydemeo commented on storage_service/locations/views.py:
```diff
@@ -44,9 +45,49 @@ def package_list(request):
+    if request.method == 'GET' and utils.get_setting('recover_backup_path_within_location') == None:
+        messages.error(request, 'Recover backup path not set.')
+
+    def execution_logic(aip): 
```
I like this refactoring of the base request logic!

mistydemeo commented on storage_service/storage_service/urls.py:
```diff
@@ -63,6 +63,11 @@ def startup():
+    recovery, _ = locations_models.Location.objects.get_or_create(
+        purpose=locations_models.Location.AIP_RECOVERY,
+        space=space,
+        relative_path=os.path.join('var', 'archivematica', 'storage_service', 'recover'),
```
I'm moving away from hardcoding these paths; see #18. (This will create a new default space with the hardcoded path if the user changes the path of the existing one.) Mind changing it to be like the ones in that other PR?

mistydemeo commented:
How do the good files make it into the AIP recover path? Is that manual?

mcantelon commented:
Yeah, good files have to be manually copied to the recover path.

mistydemeo commented on storage_service/locations/api/resources.py:
```diff
@@ -447,35 +448,35 @@ def delete_aip_request(self, request, bundle, **kwargs):
+                "message": "Deletes not allowed on this package type."
+            }
+            response_json = json.dumps(response)
+            return http.HttpMethodNotAllowed(response_json)
```
Can you specify the `content_type` too?

mistydemeo commented on storage_service/locations/views.py:
```diff
@@ -44,9 +45,43 @@ def package_list(request):
+def aip_recover_request(request):
+    def execution_logic(aip): 
+        recover_location = LocationPipeline.objects.get(
+            pipeline=aip.origin_pipeline, location__purpose='AR').location
```
Just to double-check, do we have a constraint that prevents more than one recovery location being set for a pipeline?

mcantelon replied:
Hrm, I don't think we do... I'll have to take a look at that.

mistydemeo replied:
Okay! This'll raise when the `.get` occurs, which would be confusing/inconvenient, so it seems like it'd make sense to avoid the multiple locations happening in the first place.

mistydemeo replied:
(Raise if there's more or less than one recovery location in the space, that is.)

mcantelon replied:
Okay, I've added some logic in the admin interface to prevent creation of more than one recover location per pipeline.

mistydemeo commented:
:thumbsup:

mistydemeo commented:
Oops, just realized that I accidentally closed this because I deleted the branch it's branched from.

mistydemeo commented:
Recreated the branch.

mcantelon committed:
AIP request optional remote endpoint notify.

Added configuration options/logic to enable remote REST
endpoints to be notified upon approval/rejection/failure of a
requested AIP operation. Enabled for AIP recovery.

mistydemeo commented on storage_service/locations/api/resources.py:
```diff
@@ -449,40 +450,35 @@ def delete_aip_request(self, request, bundle, **kwargs):
-            status_code = 202
+            return http.HttpMethodNotAllowed(response_json, content_type='application/json')
 
-            # This isn't configured by default
```
This whole block from the AIP deletion request path should be moved to the new AIP deletion request instead of deleted.

mcantelon committed:
AIP recovery: add Sphinx doc of admin/process

mistydemeo commented:
Added a new commit to fix an issue that can occurred when the AIP recovery location was altered from the defaults.
